tests/lp_hash_table_test.cpp

```diff
@@ -0,0 +1,57 @@
+#include "./lp_hash_table_test.h"
+#include "lp_hash_table.h"
+#include <assert.h>
+#include <iostream>
+
+class test_type {
+public:
+  int val;
+
+  test_type(int v) { this->val = v; }
+};
+
+bool test_hash_table() {
+  lp_hash_table<test_type *> table = lp_hash_table<test_type *>(1);
+
+  test_type *t = new test_type(1);
+  table.insert("key", t);
+
+  test_type *r;
+  r = table.find("key");
```

Comment:
That's not correct. Remember that we're hashing with _pointers_ (i.e. the keys are pointers). So, to hash and get the same value, we have to pass the same pointer (this is the whole point of this hash table :) ).
However, `"key"` in `insert` and `"key"` here are guaranteed to be in the same memory (i.e. same pointer).

Funny that if you ran this in optimized mode, then the compiler does string interning (which you might have seen around as another issue) on the string literals (meaning not while processing them, as we will do, but when it generates them).
That is, if 2 string literals in 2 (or more) separate locations are the same, it will save them only once and so here,
you will actually pass the same pointer.

Reply:
Same problem in the other tests :/ Which otherwise are good.

Reply:
In other words this test should fail if the implementation was indeed correct right?

Reply:
Well, except if the compiler decided to do string interning in unoptimized mode. I don't remember the last time I checked this, I'll come back. Otherwise yes, it should fail but I haven't run this.

Reply:
Ok the way I'm hashing seems right, I'm using the pointer so the 2 "key"'s should be different unless of course interning was used. I'm compiling with -Wall and -g if that helps

Reply:
Well, last time I checked GCC did not do string interning. But now it seems it does, at least for trivial cases. One test you can run yourself:
```
#include <stdio.h>
void test(const char *s) {
  printf("%p\n", s);
  ++s;
}
int main(void) {
  test("key");
  printf("whatever message\n");
  test("key");
}
```
So, even with `-g`, this will intern `"key"` and that's why your code works.

Reply:
Ok good that means that the hash function is working properly and that I just need to change the tests so they pass with interning turned off.